### PR TITLE
Fix nested type handling

### PR DESCRIPTION
**Pull request by victor-pavlychko**
First of all, thanks for a great tool! 🙂

While running swift-doc with one of my current projects, I noticed a bug with nested type handling.

Consider the following input:

```swift
public class Outer { }

extension Outer {
    public enum Inner {
        case Case
    }
}

extension Outer.Inner {
    public static let Ver = 0
}
```

This snippet triggers the following issues:
- `Case` and `Ver` symbols are mistakenly associated with `Outer` symbol
- No "nested types" and "member of" sections are generated

### CODE REVIEW

**Comment by mattt** [comment]
Thanks so much for taking the time to identify this issue and provide a fix, @victor-pavlychko. I really appreciate it!

I'd been holding off on investing in a robust test suite for this functionality, because I thought it might soon be replaced (see #31). But now that I'm more confident that the current, syntactic approach will be sticking around, I'd like to add more unit tests to exercise these kinds of corner cases.

Could you please add a test case to verify the behavior of your patch? The example you used here is great; just a few requested changes to match the conventions of the existing test:

```swift
public class C { }

extension C {
    public enum E {
        case c
    }
}

extension C.E {
    public static let tp = 0
}
```

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift [hunk 1]
```diff
@@ -73,30 +73,36 @@ struct Relationships: Component {
     }
 
     var html: HypertextLiteral.HTML {
-        var graph = symbol.graph(in: module)
-        guard !graph.edges.isEmpty else { return "" }
-
-        graph.aspectRatio = 0.125
-        graph.center = true
-        graph.overlap = "compress"
+        guard sections.map(\.symbols.count).reduce(0, +) != 0 else { return "" }
```
Comment (mattt) [inline Sources/swift-doc/Supporting Types/Components/Relationships.swift:76]: I'm still on the fence about Swift's new keypath `map` syntax. But setting that aside: We don't need to compute the total count to determine if there are any symbols. A better way would be to do `let _ = sections.first { !$0.symbols.isEmpty }`.

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift [hunk 2]
```diff
@@ -73,30 +73,36 @@ struct Relationships: Component {
...
 
-        do {
-            svg = try HypertextLiteral.HTML(String(data: graph.render(using: algorithm, to: .svg), encoding: .utf8) ?? "")
-        } catch {
-            logger.error("\(error)")
+            do {
+                svg = try HypertextLiteral.HTML(String(data: graph.render(using: algorithm, to: .svg), encoding: .utf8) ?? "")
+            } catch {
+                logger.error("\(error)")
+            }
```
Comment (mattt) [inline Sources/swift-doc/Supporting Types/Components/Relationships.swift:91]: I think this has gotten complex enough to warrant extraction into a separate helper method / property (e.g. `var svg: HyperTextLiteral.HTML`. What do you think?

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift [hunk 3]
```diff
@@ -73,30 +73,36 @@ struct Relationships: Component {
...
         }
 
         return #"""
         <section id="relationships">
             <h2 hidden>Relationships</h2>
-            <figure>
-                \#(svg ?? "")
+                \#(svg.flatMap { svg in
```
Comment (mattt) [inline Sources/swift-doc/Supporting Types/Components/Relationships.swift:97]: I'm not sure what `flatMap` is providing here as compared to the original, but that could just be me having trouble reading the diff.

File: Sources/SwiftDoc/Interface.swift [hunk 4]
```diff
@@ -55,9 +55,9 @@ public final class Interface: Codable {
     public private(set) lazy var relationships: [Relationship] = {
         var relationships: Set<Relationship> = []
         for symbol in symbols {
-            let `extension` = symbol.context.compactMap({ $0 as? Extension }).first
+            let owner = symbol.context.last(where: { $0 is Extension || $0 is Symbol })
```
Comment (mattt) [inline Sources/SwiftDoc/Interface.swift:58]: I'm not sure `owner` the right name for this. Can we come up with something better? (If not, we can go with something inelegant but precise, like `lastExtensionOrSymbolContext`)

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift (hunk 1 above)
Comment (victor-pavlychko) [reply Sources/swift-doc/Supporting Types/Components/Relationships.swift:76]: I'm not sure about checking the first item only:
- I noticed that `sections` are processed with `filter { !$0.symbols.isEmpty }` so there should be no empty sections at the moment, we can just test for `sections.isEmpty`
- at the same time, I see `guard !symbols.isEmpty else { return nil }` in the generation code

So I decided to write a more defensive check to be sure that it won't break in the future.

I'd prefer to state that we never have empty sections and simplify/remove all relevant checks.

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift (hunk 3 above)
Comment (victor-pavlychko) [reply Sources/swift-doc/Supporting Types/Components/Relationships.swift:97]: I moved the "<figure>" block inside `flatMap` so it is not rendered when the `svg` data is missing.

File: Sources/SwiftDoc/Interface.swift (hunk 4 above)
Comment (victor-pavlychko) [reply Sources/SwiftDoc/Interface.swift:58]: I can also propose `lastDeclarationScope`, what do you think?

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift (hunk 2 above)
Comment (victor-pavlychko) [reply Sources/swift-doc/Supporting Types/Components/Relationships.swift]: Agree, moved to `graphHTML` property.

**Comment by victor-pavlychko** [comment]
Thanks for the quick feedback. I have added a test for `SwiftDoc` library functionality. As for the `swift-doc` rendering code, the tests are also there, but, I have disabled them for now. It looks like SPM can't test executable targets.

**Comment by kean** [comment]
@victor-pavlychko, I was going to work on the same thing but you beat me to it 😭

> As for the swift-doc rendering code, the tests are also there, but, I have disabled them for now. It looks like SPM can't test executable targets.

I was wondering the same thing. Regardless of whether it can or can't test executables, it might be worth moving all of this functionality to SwiftDoc library.

File: Sources/SwiftDoc/Interface.swift (hunk 4 above)
Comment (mattt) [reply Sources/SwiftDoc/Interface.swift:58]: Perfect!

**Comment by kean** [comment]
I performed manual testing, the changes in this branch fixed the issue https://github.com/SwiftDocOrg/swift-doc/issues/25

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift (hunk 1 above)
Comment (mattt) [reply Sources/swift-doc/Supporting Types/Components/Relationships.swift]: To clarify: the use of `first` doesn't check the first item, but returns as soon as it finds a section with non-empty `symbols`. But taking another look, you're right: a simple check on `sections.isEmpty` should be sufficient. The other guard statement you mentioned was from an earlier iteration, and I neglected to remove it.

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift (hunk 3 above)
Comment (mattt) [reply Sources/swift-doc/Supporting Types/Components/Relationships.swift]: Ah, got it. Thanks for clarifying.

**Comment by mattt** [comment]
> @kean: Regardless of whether it can or can't test executables, it might be worth moving all of this functionality to SwiftDoc library.

I'm still not 100% on the current shape and spelling of the SwiftDoc module, but I'm pretty confident that anything to do with rendering pages should be in a separate module.

While we're waiting on end-to-end tests for the executable, could we write tests that `@testable import swift-doc`?  (I don't think I've ever tried to do anything with the module of an executable, so I'm genuinely curious). If not, then I think the next best thing for now would be to move everything but the commands themselves into a new module, and test the rendering behavior through that directly.

**Comment by victor-pavlychko** [comment]
> @mattt: While we're waiting on end-to-end tests for the executable, could we write tests that `@testable import swift-doc`? (I don't think I've ever tried to do anything with the module of an executable, so I'm genuinely curious).

Tried that: it complies swift files and then fails to link with the `swift-doc` module throwing a bunch of unresolved external errors.

I suggest leaving those test methods disabled for now and enabling them once the functionality is moved somewhere.

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift (hunk 1 above)
Comment (victor-pavlychko) [reply Sources/swift-doc/Supporting Types/Components/Relationships.swift]: Oh, right, I've misunderstood your initial suggestion. Thanks for clarifying that.

Changing to `isEmpty` + removed that `guard` statement.

File: Tests/SwiftDocTests/NestedTypesTests.swift [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        XCTAssert(`enum`.api is Enumeration)
+
+        let classRelationships = Relationships(of: `class`, in: module)
+        XCTAssertNotEqual(classRelationships.html, "")
+
+        let enumRelationships = Relationships(of: `enum`, in: module)
+        XCTAssertNotEqual(enumRelationships.html, "")
+    }
+
+    func testNoRelationshipsSeciont() throws {
```
Comment (mattt) [inline Tests/SwiftDocTests/NestedTypesTests.swift:98]:
```suggestion
    func testNoRelationshipsSection() throws {
```

File: Sources/swift-doc/Supporting Types/Components/Relationships.swift [hunk 6]
```diff
@@ -36,6 +36,23 @@ struct Relationships: Component {
         self.inheritedTypes = module.interface.typesInherited(by: symbol) + module.interface.typesConformed(by: symbol)
     }
 
+    var graphHTML: HypertextLiteral.HTML? {
```
Comment (mattt) [inline Sources/swift-doc/Supporting Types/Components/Relationships.swift:39]: Thanks for extracting this into a property. FYI: I'll probably end up further extracting this (along with the enclosing `<figure>` tag) into a new component as part of #67.

File: Tests/SwiftDocTests/NestedTypesTests.swift (hunk 5 above)
Comment (victor-pavlychko) [reply Tests/SwiftDocTests/NestedTypesTests.swift:98]: Thanks 🤦‍♂️

**Comment by mattt** [comment]
Perfect. Thanks again, @victor-pavlychko! 🎉
